packages/skeleton/src/lib/index.ts

```diff
@@ -14,6 +15,11 @@ export type { PopupSettings } from './utilities/Popup/types';
 // This type alias is to identify CSS classes within component props, which enables Tailwind IntelliSense
 export type CssClasses = string;
 
+export interface TransitionSettings {
+	transition: (...args: Parameters<typeof slide>) => TransitionConfig;
```

Comment:
@ryceg I like that @Mahmoud-zino has created a global shared type for reuse, but I'v noted we're using `slide` here as the default. Is there a more generic value we can use here? I believe transitions are just functions by default - do we provide a generic function here?

This may tie into another issue I'll raise in the PR thread in regards to removing animations.

Reply:
@endigo9740 yeah I looked into it, this will work with most transitions but not `draw` or `crossfade` because they take different parameters. I think the only way to make it really generic is by using `any`...

Reply:
We can use a union type;
 
```ts
transition: (...args: Parameters<typeof blur | fade | fly | slide | scale>) => TransitionConfig
```

It's not pretty, and it doesn't support custom transitions that include additional params (although you could slap a ` | Record<string | any>` on the end to accommodate that) but it does work

Reply:
@endigo9740 @ryceg Is using `object` a bad idea? I achieved the same functionality by changing the params to object and it works.
```ts
export interface TransitionSettings {
	transition: (node: HTMLElement, params: object) => TransitionConfig;
	params: object;
}
```

Reply:
Typescript works by narrowing wider types; the `object` type simply tells Typescript that it's not one of the other primitives. Both are technically "correct", but you're going to be more _accurate_ with a union type (accuracy in the sense of trying to make your types resemble the corresponding data). 

Try it out- with the union type you'll get intellisense on the object, and it'll narrow down the possible choices as you add props that don't exist on some of the interfaces.

Reply:
@endigo9740 @ryceg 
This is from Svelte\types... and they are using `any` to define the params.
```ts
declare type TransitionFn = (node: Element, params: any, options: TransitionOptions) => TransitionConfig;
export declare function create_in_transition(node: Element & ElementCSSInlineStyle, fn: TransitionFn, params: any): {
    start(): void;
    invalidate(): void;
    end(): void;
};
```
regarding the Union, it works perfectly as long as the user provides params from one of the transitions (which I really think are more than enough).
I tried this after @ryceg mentioned the benefits of Union and it works really well.
```ts
export type TransitionParams = BlurParams | FadeParams | FlyParams | SlideParams | ScaleParams;
export interface TransitionSettings {
	transition: (node: HTMLElement, params: TransitionParams ) => TransitionConfig;
	params: TransitionParams;
}
```
I just tested it with Svelte's custom spinning transition from the [tutorial](https://learn.svelte.dev/tutorial/custom-css-transitions).

Reply:
Rhys will roll up a newspaper and bop us on the head if we use `any` :P I'll let him comment on the param type though.

Another thought per disabling transitions - one of the primary reasons I wanted to move forward with user-provided transitions was providing a way not to just make the transition have a duration of 0, but to remove the runtime cost of animations, for performance.

The perfect example is the autocomplete component - some folks are adding hundreds or thousands of options to that, and the calculations for the transitions are running per each row. I ended up dropping animations for this component because of this. How do we cover this use case? Do we need to be able to inform users how to essentially remove the transition calculation through the passed value.

Thoughts on how we do that?

Reply:
Since the users in these specific cases have to pass undefined to both In and Out transitions separately which IMO doesn't look good, I suggest creating an extra Prop to disable the transitions.
and we can benefit from that by not reducing the duration to 0 if `$prefersReducedMotion` was set, but really just disabling the transitions. This is a lot easier to understand and I think will be cleaner in code.

Reply:
@Mahmoud-zino I'm not sure another prop would help since the transition directives are still applied to the elements. We need to alter the value passed to the `transition:in` and `transition:out`, correct?

Reply:
yes and no, we can do something like [this repl](https://svelte.dev/repl/fa6cdc5ef80d4be29b35c112a1b9c659?version=3.28.0).

Reply:
Hmm, I like it, but there's a lot of boilerplate. We definitely don't want that to live within every component. It would need to be abstracted for reuse by all components.

I'm was thinking something more like how we handle dynamic actions:
https://github.com/skeletonlabs/skeleton/blob/dev/packages/skeleton/src/lib/components/Avatar/Avatar.svelte#L20

If the action is not present, we still need to pass something to `use:` directive, so we pass an empty function:
https://github.com/skeletonlabs/skeleton/blob/dev/packages/skeleton/src/lib/components/Avatar/Avatar.svelte#L59

In practice that would look something like this:

```html
<SomeComponent
    transitionIn={{transition: () => {}, params: {}}}
    transitionOut={{transition: () => {}, params: {}}}
/>
```

I don't like this though, we're then forcing end users to pass a bunch of extra noise.

This makes me wonder if we build in and include our own custom transaction called `disabled` doesn't doesn't do anything:

```js
function disabled(node, { duration }) {
    return { duration: 0, css: t => `` };
}
```

Then allow users to provide this like so:

```html
<SomeComponent
    transitionIn={{transition: disabled, params: {}}}
    transitionOut={{transition: disabled, params: {}}}
/>
```

Then if we can make params optional, it could be:

```html
<SomeComponent
    transitionIn={{transition: disabled}}
    transitionOut={{transition: disabled}}
/>
```

But I'm not sure this guarantees there won't be a performance hit, even if it's minor.

Really stumped on this one, it's a tough one to figure out.

Reply:
@endigo9740 
I like the disabled Idea, but we don't have to return anything from it. Svelte only cares about the function parameters having the correct type.
I don't know how yet but I will try to benchmark both approaches and check the performance.
This might take some time, but I will report back when I have something.

Reply:
I'm not a huge fan of `disabled` being a function- traditionally `disabled` is a boolean, and there's less friction if we don't require them to import the `disabled` function just to pass it in. We could always add in a `disabled` flag in an extended interface very easily, and this would also afford the ability to toggle it.